### PR TITLE
Planned Procedures

### LINKED ISSUE #1110: Add Planned Procedures to eCR Viewer
## Action Requested
Add the Planned Procedures field to the eCR Viewer.
>eCR Viewer
>> Clinical Info
>>> Treatment Details
>>>> Planned Procedures

## Acceptance Criteria
Planned Procedures are displayed for eCRs that have the field.

## Testing
Unit testing will be required.

## Background & Context
This field may be a list of sub-fields.

## Designs
[Figma Designs](https://www.figma.com/file/tXWUeV3Kjehd7VNqHojZYq/eCR-Viewer?type=design&node-id=1024%3A136062&mode=design&t=lClwL7CnvrviunUF-1)
[Figma Clickable Prototype](https://www.figma.com/proto/tXWUeV3Kjehd7VNqHojZYq/eCR-Viewer?page-id=871%3A2283&type=design&node-id=1606-27619&viewport=1011%2C558%2C0.05&t=sGUJKIN0MXVFYIfm-1&scaling=min-zoom&starting-point-node-id=1606%3A27619&mode=design)

## Additional Information
We'll need to test to see whether or not this field is already being converted to FHIR, or if we'll need to complete both FHIR conversion and adding to the eCR Viewer.

Planned Procedures may be a list, so if it seems difficult to incorporate all available information into the design, please chat with Sarah.

### PR DESCRIPTION
**Pull request by JNygaard-Skylight**
# PULL REQUEST

## Summary
Add planned procedure table to eCR Viewer

## Related Issue
Fixes #1110 

## Additional Information
Anything else the review team should know?

## Checklist

- [ ] If this code affects the other scrum team, have they been notified? (In Slack, as reviewers, etc.)

[//]: # (PR title: Remember to name your PR descriptively!)

### CODE REVIEW

**Comment by JNygaard-Skylight** [comment]
@sarahtress This is ready for design review. As a note we only have one (1) example eCR with planned procedures, so if the data looks off that's why.
Screenshot 2024-04-12 at 12 46 03

**Comment by sarahtress** [comment]
@JNygaard-Skylight this looks good to me! one question though - I notice the column are different for procedures and planned procedures, is that just because they have different data available for them? I'm just curious as to why they're different

**Comment by BobanL** [comment]
> @JNygaard-Skylight this looks good to me! one question though - I notice the column are different for procedures and planned procedures, is that just because they have different data available for them? I'm just curious as to why they're different

We only have one example bundle and this is the only data inside (Procedure Name, Ordered Date, Scheduled date).

File: containers/ecr-viewer/src/app/utils.tsx [hunk 1]
```diff
@@ -795,6 +795,55 @@ export const returnProceduresTable = (
...
+    return undefined;
+  }
+
+  const columnInfo: ColumnInfoInput[] = [
+    { columnName: "Procedure Name", infoPath: "plannedProcedureName" },
+    { columnName: "Ordered Date", infoPath: "plannedProcedureOrderedDate" },
+    { columnName: "Scheduled Date", infoPath: "plannedProcedureScheduledDate" },
+  ];
+
+  carePlanActivities.forEach((entry) => {
```
Comment (KennethSkylight) [inline containers/ecr-viewer/src/app/utils.tsx:821]: Can we add a comment what this foreach is trying to accomplish?

Comment (KennethSkylight) [inline containers/ecr-viewer/src/app/utils.tsx:821]: Can we add a comment what this foreach is trying to accomplish?

File: containers/ecr-viewer/src/app/utils.tsx [hunk 2]
```diff
@@ -795,6 +795,55 @@ export const returnProceduresTable = (
...
+    if (entry.extension) {
+      const orderedDateIndex = entry.extension.findIndex(
+        (extension) => extension.url === "dibbs.orderedDate",
+      );
+
+      if (orderedDateIndex !== -1) {
```
Comment (KennethSkylight) [inline containers/ecr-viewer/src/app/utils.tsx:827]: This line also is confusing. A comment would be helpful or rewriting it as a helper function.

**Review by KennethSkylight — COMMENTED** [review]
Just the two comments added or reworks will be helpful in future maintaining of the code.

File: containers/ecr-viewer/src/app/utils.tsx (hunk 1 above)
Comment (BobanL) [reply containers/ecr-viewer/src/app/utils.tsx]: Good callout! This for each was finding the dibbs.orderedDate extension and modifying that value to be formatted. The first step here was to find the index of `dibbs.orderedDate` (-1 if not found) and then format it.

 I think i have simplified this logic by adding a new parameter to the columnInfo to apply a function to the value that is found.

File: containers/ecr-viewer/src/app/view-data/components/tests/Clinical.test.tsx [hunk 3]
```diff
@@ -344,7 +344,7 @@ describe("Check that Clinical Info components render given FHIR bundle", () => {
     // Ensure the three tables (Immunization History & Active Problems & Treatment) are rendering
```
Comment (angelathe) [inline containers/ecr-viewer/src/app/view-data/components/tests/Clinical.test.tsx:344]: Can we change these comments to match the updated tests?

`// Ensure only one table (Treatment) is rendering`
`// Ensure the three tables (Immunization History & Active Problems & Treatment) are rendering`

Comment (BobanL) [reply containers/ecr-viewer/src/app/view-data/components/tests/Clinical.test.tsx]: I'm going to just remove this comment!

File: containers/ecr-viewer/src/app/tests/evaluate-service.test.tsx [hunk 4]
```diff
@@ -241,6 +241,105 @@ describe("evaluate value", () => {
...
+        [{}],
+        mappings,
+        [{ columnName: "Col1", value: "Data1" }],
+        "",
+      ),
+    );
+
+    expect(screen.getByText("Col1")).toBeInTheDocument();
+    expect(screen.getByText("Data1")).toBeInTheDocument();
+  });
```
Comment (angelathe) [inline containers/ecr-viewer/src/app/tests/evaluate-service.test.tsx:276]: Is this a duplicate?

File: containers/ecr-viewer/src/app/tests/evaluate-service.test.tsx [hunk 5]
```diff
@@ -241,6 +241,105 @@ describe("evaluate value", () => {
...
+        [{ id: "id1" }],
+        { getId: "id" },
+        [{ columnName: "Col1", infoPath: "getId" }],
+        "",
+      ),
+    );
+
+    expect(screen.getByText("Col1")).toBeInTheDocument();
+    expect(screen.getByText("id1")).toBeInTheDocument();
+  });
```
Comment (angelathe) [inline containers/ecr-viewer/src/app/tests/evaluate-service.test.tsx:302]: Is this also a duplicate? (although lines 295 and 282 differ slightly)

File: containers/fhir-converter/Templates/eCR/Resource/_CarePlan.liquid [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        "id":"{{ID}}",
+        "identifier":
+        [
+            {% assign ids = carePlan.id | to_array -%}
+            {% for id in ids -%}
+                { {% include 'DataType/Identifier' Identifier: id -%} }
+            {% endfor -%}
+        ],
+        "status": "unknown",
+        "intent": "proposal",
```
Comment (angelathe) [inline containers/fhir-converter/Templates/eCR/Resource/_CarePlan.liquid:14]: Are these meant to be static?

File: containers/fhir-converter/Templates/eCR/DataType/_CodeableConcept.liquid [hunk 7]
```diff
@@ -1,4 +1,5 @@
 "coding": 
 [
 	{ {% include 'DataType/Coding' Coding: CodeableConcept -%} },
-],
\ No newline at end of file
+],
+"text": "{{ Text }}",
```
Comment (angelathe) [inline containers/fhir-converter/Templates/eCR/DataType/_CodeableConcept.liquid:5]: Having trouble tracing where `Text` comes from, can you give context / the code snippet? 😅

File: containers/fhir-converter/Templates/eCR/Resource/_CarePlan.liquid (hunk 6 above)
Comment (BobanL) [reply containers/fhir-converter/Templates/eCR/Resource/_CarePlan.liquid:14]: Yes - they are required in the fhir spec, but the messages we get do not have that information.
image

File: containers/ecr-viewer/src/app/utils.tsx [hunk 8]
```diff
@@ -50,8 +51,8 @@ export interface ColumnInfoInput {
   columnName: string;
   infoPath?: string;
   value?: string;
-  sentenceCase?: boolean;
   hiddenBaseText?: string;
+  applyToValue?: (value: any) => any;
```
Comment (angelathe) [inline containers/ecr-viewer/src/app/utils.tsx:55]: Thanks for adding this!! 🎉

File: containers/fhir-converter/Templates/eCR/DataType/_CodeableConcept.liquid (hunk 7 above)
Comment (BobanL) [reply containers/fhir-converter/Templates/eCR/DataType/_CodeableConcept.liquid:5]: Good catch - we used this previously, but then stopped but stayed in. I'll remove it.

Comment (angelathe) [reply containers/fhir-converter/Templates/eCR/DataType/_CodeableConcept.liquid:5]: @BobanL Ok sweet! I'll approve once this gets removed

**Review by angelathe — APPROVED** [review]

**Pull request merged**